### DISCUSSION TITLE
is there a DJL best practice？

### DISCUSSION THREAD

**Discussion by iDonal** [pyg-team/pytorch_geometric#4537] [Q&A]
because of torch-scatter and torch-sprase 
DJL（https://github.com/deepjavalibrary/djl） cant use pre-build pytorch-lib
we should build pytorch_geometric and jni manually

**Comment by rusty1s** [comment]
Can you share some more information on this? I do not have any experience with `djl` yet.

**Reply by iDonal** [reply]
Djl can load multiple popular DL models： like MXNET TENSOR_FLOW PYTORCH  using java language 
For pytorch， the calling sequence likes ：  java->jni->torch c++ front head
I successfully load torch_geometric model by torch c++ front head  , but djl failed

**Reply by iDonal** [reply]
Exception:
Exception in thread "main" java.lang.RuntimeException: ai.djl.engine.EngineException: 
Unknown builtin op: torch_sparse::ptr2ind.
Could not find any similar ops to torch_sparse::ptr2ind. This op may not exist or may not be currently supported in TorchScript.

**Reply by iDonal** [reply]
DJL official pytorch example: https://github.com/deepjavalibrary/djl/blob/master/jupyter/pytorch/load_your_own_pytorch_bert.ipynb

**Reply by rusty1s** [reply]
Have you seen https://github.com/rusty1s/pytorch_scatter#c-api? Any chance one can register a custom C++ library into `djl`? Other-wise I am not sure how this should be fixable :(

**Reply by iDonal** [reply]
Thanks!  I try to learn something about JNI first.   C++ example  works.
